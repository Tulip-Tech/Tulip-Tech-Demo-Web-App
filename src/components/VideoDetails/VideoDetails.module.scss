@use 'src/styles/variables';
@use 'src/styles/theme';
@use 'src/styles/mixins/responsive';
@use 'src/styles/mixins/typography';

.video {
  color: var(--primary-color);
  font-family: var(--body-font-family);
  font-weight: 400;
  text-shadow: 0 2px 4px rgba(0, 0, 0, 0.14), 0 3px 4px rgba(0, 0, 0, 0.12), 0 1px 5px rgba(0, 0, 0, 0.2);
}

.mainPadding {
  padding: 37px 56px 0;

  @include responsive.desktop-small-only() {
    padding: 34px 36px 36px;
  }

  @include responsive.tablet-only() {
    padding: 34px 24px 24px;
  }

  @include responsive.mobile-only() {
    padding: 16px;
  }
}

.main {
  display: flex;
  max-width: 100vw;

  &.hidden {
    display: none;
  }
}

.info {
  position: relative;
  width: 70%;
  max-width: 650px;
  min-height: 440px;

  @include responsive.desktop-small-only() {
    min-height: 335px;
  }

  @include responsive.tablet-only() {
    width: 350px;
    min-height: 335px;
  }

  @include responsive.mobile-only() {
    width: 100%;
    min-height: 380px;
    padding-top: 225px;
  }
}

.title {
  @include typography.video-title-base;
  margin-bottom: calc(#{variables.$base-spacing} / 2);
}

.metaContainer {
  display: flex;
  flex-direction: column;
}

.primaryMetadata {
  margin-bottom: 8px;
  font-size: 16px;
  line-height: 18px;
  letter-spacing: 0.15px;

  @include responsive.mobile-only() {
    order: 2;
    font-size: 14px;
  }
}

.secondaryMetadata {
  margin-top: 24px;
  margin-bottom: 8px;
  font-size: 20px;
  line-height: 20px;
  letter-spacing: 0.5px;

  @include responsive.mobile-only() {
    order: 1;
    margin: 4px 0;
    font-size: 18px;
  }
}

.description {
  min-height: 21px;
  margin-bottom: 24px;
  font-size: 18px;
  line-height: 20px;
  letter-spacing: 0.5px;
}

.buttonBar {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  margin: 0 -6px;

  > button {
    justify-content: center;
    margin: 6px;
  }

  @include responsive.mobile-and-tablet() {
    flex-wrap: wrap;

    > button {
      flex: 1;
      padding: 0 12px;

      &:first-child {
        flex-basis: 100%;
      }
    }
  }

  @include responsive.mobile-only() {
    flex-wrap: wrap;

    > button:first-child {
      margin-bottom: 8px;
    }
  }
}

.poster {
  position: absolute;
  top: 0;
  right: 0;

  z-index: -1;

  width: 80vw;
  height: calc(85vw / 16 * 9);

  background-repeat: no-repeat;
  background-position: center center;
  background-size: cover;
  // fix for Safari (poster will render below the dynamic blur)
  transform: translateZ(0);

  mask-image: radial-gradient(farthest-corner at 80% 20%, rgba(0, 0, 0, 1) 30%, rgba(0, 0, 0, 0) 60%);
  //noinspection CssInvalidPropertyValue
  -webkit-mask-image: radial-gradient(farthest-corner at 80% 20%, rgba(0, 0, 0, 1) 30%, rgba(0, 0, 0, 0) 60%); /* stylelint-disable-line */

  @include responsive.tablet-only() {
    width: 80vw;
    height: calc(140vw / 16 * 9);
  }

  @include responsive.mobile-only() {
    width: 110vw;
    height: calc(180vw / 16 * 9);
  }
}

.playIcon {
  position: absolute;
  top: calc(50% - 60px);
  left: calc(50% - 35px);
  padding: 25px 30px;
  font-size: xx-large;
  border: 3px solid white;
  border-radius: 100%;
}
