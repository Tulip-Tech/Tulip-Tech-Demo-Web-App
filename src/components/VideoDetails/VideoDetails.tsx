import React from 'react';
import classNames from 'classnames';

import styles from './VideoDetails.module.scss';

import CollapsibleText from '#components/CollapsibleText/CollapsibleText';
import useBreakpoint, { Breakpoint } from '#src/hooks/useBreakpoint';
import Image from '#components/Image/Image';
import type { ImageData } from '#types/playlist';
import { testId } from '#src/utils/common';

type Props = {
  title: string;
  description: string;
  primaryMetadata: React.ReactNode;
  secondaryMetadata?: React.ReactNode;
  image?: ImageData;
  startWatchingButton: React.ReactNode;
  shareButton: React.ReactNode;
  favoriteButton: React.ReactNode;
  trailerButton: React.ReactNode;
};

const VideoDetails: React.VFC<Props> = ({
  title,
  description,
  primaryMetadata,
  secondaryMetadata,
  image,
  startWatchingButton,
  shareButton,
  favoriteButton,
  trailerButton,
}) => {
  const breakpoint: Breakpoint = useBreakpoint();
  const isMobile = breakpoint === Breakpoint.xs;

  return (
    <div className={styles.video} data-testid={testId('video-details')}>
      <div className={classNames(styles.main, styles.mainPadding)}>
        <Image className={styles.poster} image={image} alt={title} width={1280} />
        <div className={styles.info}>
          <h2 className={styles.title}>{title}</h2>
          <div className={styles.metaContainer}>
            <div className={styles.primaryMetadata}>{primaryMetadata}</div>
            {secondaryMetadata && <div className={styles.secondaryMetadata}>{secondaryMetadata}</div>}
          </div>
          <CollapsibleText text={description} className={styles.description} maxHeight={isMobile ? 60 : 'none'} />

          <div className={styles.buttonBar}>
            {startWatchingButton}
            {trailerButton}
            {favoriteButton}
            {shareButton}
          </div>
        </div>
      </div>
    </div>
  );
};

export default VideoDetails;
