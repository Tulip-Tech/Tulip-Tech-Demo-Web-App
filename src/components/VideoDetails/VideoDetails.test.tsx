import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import VideoDetails from './VideoDetails';

describe('<VideoDetails>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(
      <VideoDetails
        title="Test video"
        description="Video description"
        primaryMetadata="Primary metadata string"
        secondaryMetadata={<strong>Secondary metadata string</strong>}
        image={{ image: 'http://image.jpg' }}
        startWatchingButton={<button>Start watching</button>}
        shareButton={<button>share</button>}
        favoriteButton={<button>favorite</button>}
        trailerButton={<button>play trailer</button>}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('renders the image', () => {
    const { getByAltText } = render(
      <VideoDetails
        title="Test video"
        description="Video description"
        primaryMetadata="Primary metadata string"
        secondaryMetadata={<strong>Secondary metadata string</strong>}
        image={{ image: 'http://image.jpg' }}
        startWatchingButton={<button>Start watching</button>}
        shareButton={<button>share</button>}
        favoriteButton={<button>favorite</button>}
        trailerButton={<button>play trailer</button>}
      />,
    );

    expect(getByAltText('Test video')).toHaveAttribute('src', 'http://image.jpg?width=1280');
  });

  test('renders the fallback image when the image fails to load', () => {
    const { getByAltText } = render(
      <VideoDetails
        title="Test video"
        description="Video description"
        primaryMetadata="Primary metadata string"
        secondaryMetadata={<strong>Secondary metadata string</strong>}
        image={{ image: 'http://image.jpg', fallbackImage: 'http://fallback.jpg' }}
        startWatchingButton={<button>Start watching</button>}
        shareButton={<button>share</button>}
        favoriteButton={<button>favorite</button>}
        trailerButton={<button>play trailer</button>}
      />,
    );

    expect(getByAltText('Test video')).toHaveAttribute('src', 'http://image.jpg?width=1280');

    fireEvent.error(getByAltText('Test video'));

    expect(getByAltText('Test video')).toHaveAttribute('src', 'http://fallback.jpg?width=1280');
  });
});
