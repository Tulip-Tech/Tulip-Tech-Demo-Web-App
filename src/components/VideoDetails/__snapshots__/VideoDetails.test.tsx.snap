// Vitest Snapshot v1

exports[`<VideoDetails> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_video_d0c133"
    data-testid="video-details"
  >
    <div
      class="_main_d0c133 _mainPadding_d0c133"
    >
      <img
        alt="Test video"
        class="_poster_d0c133 _image_4c41c3"
        src="http://image.jpg?width=1280"
      />
      <div
        class="_info_d0c133"
      >
        <h2
          class="_title_d0c133"
        >
          Test video
        </h2>
        <div
          class="_metaContainer_d0c133"
        >
          <div
            class="_primaryMetadata_d0c133"
          >
            Primary metadata string
          </div>
          <div
            class="_secondaryMetadata_d0c133"
          >
            <strong>
              Secondary metadata string
            </strong>
          </div>
        </div>
        <div
          class="_collapsibleText_561522"
        >
          <div
            class="_textContainer_561522 _description_d0c133"
            style="max-height: none;"
          >
            Video description
          </div>
        </div>
        <div
          class="_buttonBar_d0c133"
        >
          <button>
            Start watching
          </button>
          <button>
            play trailer
          </button>
          <button>
            favorite
          </button>
          <button>
            share
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
